man/dbrestor.8
.TH DBRESTOR 8 "12 September 1995"
.FB
.SH NAME
DBRESTOR \- restore a \fIFirstBase\fP database from disk
.SH SYNOPSIS
.B DBRESTOR
.I database
.SH DESCRIPTION
.I DBRESTOR
is a shell script that will restore a "dirty" database using the
previously dumped database and the current database log file.
.PP
The concept is as follows: the dumped database and its map are copied from
the \fIdbase.BKfirstbase\fP
and \fIdbase.BKmap\fP files back to the live files \fIdbase.cdb\fP
and \fIdbase.map\fP. Then, the \fIdbrestor\fP(8)
tool will use the database log,
\fIdbase.log\fP,
and recover the database almost to the point where the failure occurred.
.PP
Note: it is very important that the backed up database and its map are
kept intact before using the \fIdbrestor\fP(8) script. This is due to the fact
the the database log is kept using \fIlogical\fP record numbers.
\fIDbclean\fP(1) would change these record numbers \(em thus causing havoc and
destruction during the \fIdbrestor\fP(8) process.
.PP
\fIDbrestor\fP(8)
will run \fIdbclean\fP(1) on the restored database once
it has completed. This is used to determine accurate counts of valid
and deleted records.
.PP
This script could easily be changed to use tape backups if desired.
.SH FILES
.PD 0
.TP 10
*.cdb
\fIFirstBase\fP database.
.TP 10
*.log
\fIFirstBase\fP database log.
.TP 10
*.map
record map of \fIFirstBase\fP database.
.TP 10
*.BKfirstbase
dumped \fIFirstBase\fP database.
.TP 10
*.BKmap
dumped map of \fIFirstBase\fP database.
.PD
.SH SEE ALSO
cp(1), firstbase(5), dictionaries(5), dbdump(8)
.PP
.I FirstBase User's Guide and Reference Manual
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dbtools/dbrestor.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbrestor.c,v 9.1 2001/02/16 19:47:51 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Dbrestor_sid[] = "@(#) $Id: dbrestor.c,v 9.1 2001/02/16 19:47:51 john Exp $";
#endif

/*
 *  dbrestor - is used to recover a fb_database from a KNOWN good database
 *	and KNOWN good fb_database map, using the fb_database log.
 */
 
#include <fb.h>
#include <fb_vars.h>

#if !FB_PROTOTYPES
static void restor();
static void fixdb();
static int loadlog();
static int scanckp();
static int loadrn();
static int nextline();
static int loadfields();
static int nextfield();
static int nextread();
#else /* FB_PROTOTYPES */
extern main(int, char **);
static void restor(char *);
static void fixdb(void);
static int loadlog(void);
static int scanckp(void);
static int loadrn(void);
static int nextline(char *);
static int loadfields(void);
static int nextfield(char *);
static int nextread(char *);
/*static void usage(void);*/
#endif /* FB_PROTOTYPES */

extern char *cdb_DBCLEAN;
#define BLKSIZE		1024
#define SEPARATOR	'|'

fb_database *hp;			/* pointer to current database */
extern short cdb_allow_dirty;
long cpos = 0;				/* current log position */
char buffer[BLKSIZE + 1];		/* for buffered read */
int verbose = 0;

/*
 * main - mainline logic for dbrestor
 */

   main(argc, argv)
   int argc;
   char *argv[];
      {
	 char dname[FB_MAXNAME];
	 int j;

         (void) Dbrestor_sid;

         fb_getargs(argc, argv, FB_NODB);
	 if (fb_testargs(argc, argv, "-v") > 0){
	    verbose = 1;
	    cdb_batchmode = 1;
	    }
	 for (j = 1; j < argc; j++)
	    if (!(argv[j][0] == '-')){
	       fb_rootname(dname, argv[j]);
	       restor(dname);
	       }
         fb_ender();
      }
      
/*
 *  restor - simple restore of a fb_database using the dbase log.
 */      

   static void restor(dname)
      char *dname;
      
      {
         int j;
	 char msg[FB_MAXLINE], com[FB_MAXLINE];
	 
	 hp = fb_dballoc();
	 cdb_allow_dirty = 1;
 	 fb_dbargs(dname, NIL, hp);
	 fb_opendb(hp, READWRITE, FB_NOINDEX, 0);
	 fb_initlog(hp);
	 fb_scrhdr(hp, "Check Files"); fb_infoline();
	 if (!cdb_batchmode && !cdb_yesmode){
	    fb_move(4, 1);
	    fb_printw("%s%s!!!\n%s\n%s\n",
	       "Make sure no one else is using ", hp->dbase,
	       "After restoring database, all indexes for",
	       "that database will require regeneration.");
	    sprintf(msg, "Ready to Restore %s (y/<no>) ?", hp->dbase);
	    if (fb_mustbe('y', msg, cdb_t_lines, 1) != FB_AOK)
	       fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	    }
	 for (j = 0; j < hp->nfields; j++){
	    if (hp->kp[j]->aid != NULL){	/* abuse msg buffer */
	       sprintf(msg, "%s.idict", hp->kp[j]->aid->autoname);
	       unlink(msg);
	       }
	    }
	 fixdb();
	 fb_setdirty(hp, 0);
	 fb_rootname(msg, cdb_db->dbase);
	 if (!cdb_batchmode){
	    fb_move(4,1), fb_clrtobot();
	    fb_scrstat("Cleaning");
	    sprintf(com, "Cleaning %s", msg);
	    fb_fmessage(com);
	    }
	 fb_closedb(hp);
	 sprintf(com, "%s -b %s", cdb_DBCLEAN, msg);
	 fb_system(com, FB_WITHROOT);
      }

/*
 *  fixdb - fix the open fb_database using the log.
 *            good records to the new fb_database (nfd) and recording
 *            their spots in the fb_database map (mfd).
 */
 
   static void fixdb()
       
      {
         int st;
	 long entry;

	 if (!cdb_batchmode){ 
	    fb_move(4, 1); fb_clrtobot();
	    fb_scrstat("Restoring");
	    fb_infoline();
	    }
	 for (entry = 0; ; entry++){
	    st = loadlog();
	    if (st == FB_ERROR)
	       fb_xerror(FB_MESSAGE, "Error in Loading Log Record:", hp->dlog);
	    else if (st == 0)
	       break;
	    if (!cdb_batchmode || verbose){
	       if (!cdb_batchmode)
	          fb_move(12, 22);
	       fb_printw("Log Entry: %ld     Database Record: %-10ld\n",
	          entry, hp->rec);
	       }
            if (fb_putrec(hp->rec, hp)==FB_ERROR)
	       fb_xerror(FB_WRITE_ERROR, SYSMSG[S_RECORD], hp->dbase);
	    }
	 if (entry == 0)
	    fb_serror(FB_MESSAGE, "Warning: No Complete Log Records Found",NIL);
	 if (!cdb_batchmode || verbose)
	    fb_serror(FB_MESSAGE, "Database Restore Successful:", hp->dbase);
      }

/*
 * loadlog - load a complete log record
 */

   static int loadlog()
      {
         int st;

         if (lseek(hp->logfd, cpos, 0) < 0){
	    if (verbose)
	       fb_serror(FB_SEEK_ERROR, "Log File:", hp->dbase);
	    return(FB_ERROR);
	    }
	 if (scanckp() != FB_AOK)
	    return(0);		/* must be end, without a CK rec */
         if (lseek(hp->logfd, cpos, 0) < 0){
	    if (verbose)
	       fb_serror(FB_SEEK_ERROR, "Log File:", hp->dbase);
	    return(FB_ERROR);
	    }
	 st = loadrn();
	 if (st == 0)
	    return(0);
	 else if (st == FB_ERROR){
	    if (verbose)
	       fb_serror(FB_SEEK_ERROR, "Load Record Number Error:", hp->dbase);
	    return(FB_ERROR);
	    }
#if XDEBUG
         fb_printw("Loaded rec number is %ld\n", hp->rec);
#endif
	 if (loadfields() == FB_ERROR){
	    if (verbose)
	       fb_serror(FB_SEEK_ERROR, "Load Fields Error:", hp->dbase);
	    return(FB_ERROR);
	    }
	 cpos = lseek(hp->logfd, 0L, 1);
	 return(FB_AOK);
      }

/*
 * scanckp - scan forward for a check point record.
 */

   static int scanckp()
      {
         char c;
	 int state = 1;

         for (; nextread(&c) != 0; ){
	    switch(state){
	       case 1:
	          if (c == '@')
		     state = 2;
		  else
		     state = 1;
	          break;
	       case 2:
	          if (c == 'C')
		     state = 3;
		  else
		     state = 1;
	          break;
	       case 3:
	          if (c == FB_NEWLINE)
		     return(FB_AOK);
		  else
		     state = 1;
	          break;
	       }
	    }
	 return(FB_ERROR);
      }

/*
 * loadrn - load the next record number - skip any checkpoint records.
 */

   static int loadrn()
      {
         char recnum[FB_MAXLINE];
         
	 if (nextline(recnum) == 0 || recnum[0] != '@')
	    return(FB_ERROR);
	 if (equal(recnum, "@C")){
	    cpos = lseek(hp->logfd, 0L, 1);
	    if (scanckp() != FB_AOK)
	       return(0);		/* must be end, without a CK rec */
	    if (lseek(hp->logfd, cpos, 0) < 0){
	       if (verbose)
	          fb_serror(FB_SEEK_ERROR, "Log File:", hp->dbase);
	       return(FB_ERROR);
	       }
	    if (nextline(recnum) == 0)
	       return(0);
	    if (recnum[0] != '@')
	       return(FB_ERROR);
	    }
	 hp->rec = atol(recnum + 1);
	 return(FB_AOK);
      }

/*
 * nextline - get a line from fb_input, strip FB_NEWLINE, and store NULL
 */

   static nextline(buf)
      char *buf;

      {
         char c;
	 int i;
         
	 for (i = 1; i < FB_MAXLINE; i++){		/* 1..FB_MAXLINE inclusive */
	    if (nextread(&c) == 0)
	       return(0);
	    if (c == FB_NEWLINE)
	       break;
	    *buf++ = c;
	    }
	 *buf = NULL;
	 return(i);
      }

/*
 * loadfields - load all the fields of the record
 */

   static int loadfields()
      {
         char c;
	 int i;

         for (i = 0; i <= hp->nfields; i++){
	    if (hp->kp[i]->type != FB_FORMULA && hp->kp[i]->dflink == NULL){
	       if (nextfield(cdb_afld) == 0){
	          if (verbose)
		     fb_serror(FB_MESSAGE, "Nextfield Error", NIL);
		  return(FB_ERROR);
		  }
	       if (verbose)
                  fb_printw("\tField %d: %s\n", i, cdb_afld);
	       fb_store(hp->kp[i], cdb_afld, hp);
	       }
	    }
	 if (nextread(&c) == 0 || c != FB_NEWLINE){
	    if (verbose)
	       fb_serror(FB_MESSAGE, "No Newline after record", NIL);
	    return(FB_ERROR);
	    }
	 return(FB_AOK);
      }

/*
 * nextfield - get a fb_field from fb_input, strip PIPE, and store NULL
 */

   static int nextfield(buf)
      char *buf;

      {
         char c, lastc = NULL;
	 int i;
         
	 for (i = 1; i < hp->recsiz; i++){
	    if (nextread(&c) == 0)
	       return(0);
	    while (c == CHAR_BACKSLASH){
	       lastc = c;
	       if (nextread(&c) == 0)
		  return(0);
	       if (c == CHAR_BACKSLASH || (c != CHAR_BACKSLASH && c != SEPARATOR))
	          *buf++ = CHAR_BACKSLASH;
	       }
	    if (c == SEPARATOR && lastc != CHAR_BACKSLASH)
	       break;
	    *buf++ = c;
	    lastc = c;
	    }
	 *buf = NULL;
	 if (c != SEPARATOR){
	    if (verbose)
	       fb_serror(FB_MESSAGE, "No Field Seperator Found", NIL);
	    return(0);
	    }
	 return(i);
      }

/*
 * nextread - read a single character, store in buf.
 */

   static nextread(buf)
      char *buf;
      
      {
         if (read(hp->logfd, buf, 1) != 1)
	    return(0);
	 return(1);
      }
      
/*
 *  usage - for dbrestor
 */
/*
*   static void usage()
*      {
*         fb_xerror(FB_MESSAGE,
*	    "usage: dbrestor [-b] [-y] dbase [dbase ...]", NIL);
*      }
*/
